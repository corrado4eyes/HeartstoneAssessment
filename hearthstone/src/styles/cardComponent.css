.cardDetail-font {
    color: #e0e0e0;
    font-family: 'Aclonica';
    text-align: center;
}

.modalBody {
    background-image: url('../assets/card-detail-background.jpg');
}

.modalHeader {
    z-index: 1;
    background-image: url('../assets/navBar.png');
}

.cardDetailContainer {
    z-index: 2;
}