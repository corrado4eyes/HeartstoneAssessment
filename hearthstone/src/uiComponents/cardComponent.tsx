import * as React from 'react';
import { Image, Card, Button, Row, Col } from 'react-bootstrap';
import CardModel from '../model/card';
import noImg from '../assets/no-img.jpg';

interface OwnProps {
    card: CardModel;
}

interface OwnState {
    goldImg: boolean
}
export class CardComponent extends React.PureComponent<OwnProps, OwnState> {
    constructor(props: OwnProps){
        super(props)
        this.state = {
            goldImg: false
        }
    }

    switchImage = () => {
        this.setState({goldImg: !this.state.goldImg})
    }

    addToFavourites = () => {
        throw("Not Implemented yet.")
    }

    render() {
        return(
            <Card className="no-border">
                <Card.Img variant="top" src={(!this.state.goldImg ? this.props.card.img : this.props.card.imgGold) || noImg}/>
                <Card.Title>{this.props.card.name}</Card.Title>
                <Card.Text>
                {`Attack: ${this.props.card.attack || "NaN"}\n
                  Health: ${this.props.card.health || "NaN"}\n`}
                </Card.Text>
                <Row>
                    <Col>
                        <Button variant={!this.props.card.imgGold ? "danger" : "primary"} onClick={this.switchImage.bind(this)} disabled={this.props.card.img ? false : true}>Change image</Button>
                    </Col>
                    <Col>
                        <Button onClick={this.addToFavourites.bind(this)}>Add to Favourites</Button>
                    </Col>
                </Row>
            </Card>
        );
    }
}