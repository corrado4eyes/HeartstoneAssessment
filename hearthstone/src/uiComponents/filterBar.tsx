import * as React from 'react';
import Navbar from 'react-bootstrap/Navbar';
import {Image, NavDropdown, Nav, Form, Button, Row, Col} from 'react-bootstrap';
import '../styles/filterBar.css';

import logo from '../assets/hearthstone-mini-logo.png';
import shinyLogo from '../assets/hearthstone-mini-logo-shiny.png';
import filterBarImg from '../assets/navBar.png';
import { CardSet } from '../model/cardSet';
import { Rarity } from '../model/rarity';
import {dispatchFilter, dispatchFilterByName} from '../redux/actions/cardActions';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { RootState } from '../redux/reducers/mainReducer';
const filters: Filter[] = [{
    name: "Card Set",
    enum: CardSet,
    id: "cardSetFilter",
    key: "cardSet"
},
{
    name: "Rarity",
    enum: Rarity,
    id: "rarityFilter",
    key: "rarity"
}]

interface Filter {
    name: string,
    enum: any
    id: string
    key: string
}

interface StateProps {
    filters: any
}

interface ActionProps {
    dispatchFilter: typeof dispatchFilter
    dispatchFilterByName: typeof dispatchFilterByName
}

interface OwnState {
    shinyImg: boolean
}

const mapDispatchToProps = (dispatch: any) => {
    return bindActionCreators({dispatchFilter, dispatchFilterByName}, dispatch)
}
const mapStateToProps = ({card}: RootState) => {
    return {
        filters: card.filters
    }
}

type Props = ActionProps & StateProps;
export class FilterBar extends React.PureComponent<Props, OwnState> {
    constructor(props: Props) {
        super(props)
        this.state = {
            shinyImg: false
        }
    }

    onFilterSelected = (filter: string, filterKey: string) => {
        this.props.dispatchFilter(filter, filterKey);
    }

    onFilterByName = (e: any) => {
        e.preventDefault();
        const text = e.currentTarget.cardName.value
        if(text.trim() === '')
            return
        this.props.dispatchFilterByName(text);
    }

    render() {
        return(
            <Row className="justify-content-md-center" >
                <Navbar className="navbar" bg="dark" expanded expand="lg">
                    <Col>
                        <Navbar.Brand >
                            <Image 
                            src={!this.state.shinyImg ? logo : shinyLogo} 
                            onMouseEnter={() => this.setState({shinyImg: true})} 
                            onMouseLeave={() => this.setState({shinyImg: false})} 
                            width="64px" height="64px"/>
                        </Navbar.Brand>
                    </Col>
                        <Navbar.Collapse>
                            <Nav>
                                {
                                    filters.map((el, i) => {
                                        return (
                                            <Col key={i}>
                                                <NavDropdown title={this.props.filters[el.key]} id={el.id} key={el.id}>
                                            {
                                                Object.values(el.enum).map((value) => {
                                                    return <NavDropdown.Item 
                                                            key={value as string} 
                                                            onClick={this.onFilterSelected.bind(this, (value as string), el.key)}>
                                                            >
                                                                {value as string}
                                                            </NavDropdown.Item>
                                                })
                                            }
                                        </NavDropdown></Col>)
                                    })
                                }
                            </Nav>
                                <Form inline onSubmit={this.onFilterByName.bind(this)}>
                                    <Form.Group>
                                        <Col>
                                            <Form.Control name="cardName" id="nameFilter" type="text" placeholder="Search" className="mr-sm-2" />
                                        </Col>
                                        <Col>
                                            <Button type="submit" variant="outline-success">Search</Button>
                                        </Col>
                                    </Form.Group>
                                </Form>
                        </Navbar.Collapse>
                </Navbar>
            </Row>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterBar)