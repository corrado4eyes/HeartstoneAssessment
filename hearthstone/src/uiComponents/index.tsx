import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';
import "reflect-metadata";
import * as Inversify from "inversify-react";
import container from '../../config/inversify';
import { configureStore } from '../redux/store/storeConfig';
import * as Redux from "react-redux";

const store = configureStore(container);
ReactDOM.render(
    <Inversify.Provider container={container}>
        <Redux.Provider store={store}>
            <App cardSet={"Basic"}/>
        </Redux.Provider>
    </Inversify.Provider>
, document.getElementById('root'));
