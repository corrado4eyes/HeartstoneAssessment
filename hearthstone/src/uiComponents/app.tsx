import React from 'react';
import '../styles/app.css';
import Card from '../model/card';
import {Container, Row, Col, Image, Spinner} from 'react-bootstrap';
import {dispatchSyncCard} from '../redux/actions/cardActions'
import {RootState} from '../redux/reducers/mainReducer';
import { bindActionCreators } from 'redux';
import {connect} from "react-redux";
import { CardComponent } from './cardComponent';
import '../styles/commonProperties.css';
import FilterBar from './filterBar';

interface StateProps {
    filteredCards: Card[];
    loading: boolean;
}

interface ActionProps {
    dispatchSyncCard: typeof dispatchSyncCard;
}

interface OwnProps {
    cardSet: string;
}

const mapStateToProps = ({card}: RootState, ownProps: OwnProps) => {
    return{
        filteredCards: card.filteredCards,
        cardSet: ownProps.cardSet,
        loading: card.loading
    }
};

const mapDispatchToProps = (dispatch: any) => {
    return bindActionCreators({dispatchSyncCard}, dispatch);
}
type Props = StateProps & ActionProps & OwnProps;
export class App extends React.PureComponent<Props, {}> {
    
    componentDidMount(){
        this.props.dispatchSyncCard()
    }

    render(){
        const cards = this.props.filteredCards;
        return this.props.loading ? 
        (<Container >
            <Row className="justify-content-md-center">
                <Col md="auto" className="vertically-centered">
                    <Spinner animation="grow" variant="success" />
                </Col>
            </Row>
        </Container>) 
        : (<Container>
                <Row><FilterBar/></Row>
                <Row>
                    { 
                        !(cards.length === 0) ?
                            cards.map((el: Card) => {
                                return (el.img || el.imgGold) ? (
                                    <Col key={el.cardId} className='card no-border' md="4">
                                        <CardComponent card={el}/>
                                    </Col>
                                ) : undefined
                            })
                            :
                            <Row className="justify-content-md-center">
                                <Col md="auto" className="vertically-centered">
                                    No Results Found!
                                </Col>
                            </Row>
                    }
                </Row>
            </Container>);
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
