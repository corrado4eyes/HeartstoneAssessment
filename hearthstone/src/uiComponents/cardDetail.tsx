import * as React from 'react';
import { CardState,  } from './cardComponent';
import CardModel from '../model/card';
import noImg from '../assets/noImg.jpg';
import changeImg from '../assets/change-img-32x32.png';
import favoriteImg from '../assets/favorite.png';
import notFavorite from '../assets/not-favorite.png';
import manaIcon from '../assets/mana-icon.png';
import { Container, Row, Col, Image, Button } from 'react-bootstrap';

export interface CardDetailProps {
    cardState: CardState
    card: CardModel
    switchImage: () => void
    addToFavourites: () => void
}

export interface OwnState {
    show: boolean
}

export class CardDetail extends React.PureComponent<CardDetailProps, OwnState> {
    constructor(props: CardDetailProps) {
        super(props)
        this.state = {
            show: false,
        }
    }

    render() {
        return (
            <Container className="cardDetailContainer">
                <Row style={{height: "auto"}}>
                    <Col>
                        <Image src={
                        (this.props.cardState.isImgAvailable) ? 
                            ((this.props.card.img) ? 
                                (this.props.cardState.goldImg ? 
                                    this.props.card.imgGold 
                                : this.props.card.img) 
                            : noImg) 
                        : noImg
                        } height="90%" width="90%"/>
                    </Col>
                    <Col className="cardDetail-font">
                        <Col>
                            {this.props.card.name}
                        </Col>
                        <Row>
                            <Col>
                                Card set: {this.props.card.cardSet}
                            </Col>
                        </Row>
                        <Row>
                            <Col> 
                                Rarity: {this.props.card.rarity} 
                            </Col>
                        </Row>
                        <Row>
                            <Col> 
                                Type: {this.props.card.type} 
                            </Col>
                        </Row>
                        <Row>
                            <Col> 
                                Player class: {this.props.card.playerClass} 
                            </Col>
                        </Row>
                        <Row>
                            <Col> 
                                Cost: {this.props.card.cost} <Image src={manaIcon}/>
                            </Col>
                        </Row>
                        <Row>
                            <Col>{ this.props.card.attack ? `Attack: ${this.props.card.attack}\n` : null}</Col>
                            <Col>{ this.props.card.health ? `Health: ${this.props.card.health}\n`: null}</Col>
                        </Row>
                    </Col>
                </Row>
                <Row>
                    <Col>
                        <Button id="switchImg" 
                            className="no-borders" 
                            onClick={this.props.switchImage} 
                            disabled={this.props.card.imgGold ? false : true}>
                            <Image src={changeImg}/>
                        </Button>
                    </Col>
                    <Col>
                        <Button id="favBtn" 
                            onClick={this.props.addToFavourites}>
                                <Image src={this.props.card.favourite ? favoriteImg : notFavorite}/>
                        </Button>
                    </Col>
                </Row>
                </Container>
        )
    }

}