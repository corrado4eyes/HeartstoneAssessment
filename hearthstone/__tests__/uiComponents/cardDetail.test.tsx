import * as React from 'react';
import {CardDetail} from '../../src/uiComponents/cardDetail';
import { shallow, ShallowWrapper, ReactWrapper, mount } from 'enzyme';
import { dummyCard } from '../__mocks__/mockObjects';
import { CardState } from '../../src/uiComponents/cardComponent';
import { CardDetailProps } from '../../src/uiComponents/cardDetail';
import { Button } from 'react-bootstrap';

const addToFavorite = jest.fn()
const switchImage = jest.fn()

const cardStateInitialValues: CardState = {
    goldImg: false,
    isImgAvailable: true,
    isOpen: false
}


describe('CardDetail', () => {
    const component: ReactWrapper = mount(<CardDetail switchImage={switchImage} addToFavourites={addToFavorite} cardState={cardStateInitialValues} card={dummyCard}/>);

    afterAll(() => {
        component.unmount()
    })

    it('matches the snapshot', () => {
        expect(component).toMatchSnapshot();
    })

    it('checks the props', () => {
        const props: CardDetailProps = (component.find(CardDetail).props() as CardDetailProps)
        expect(props.cardState.goldImg).toBe(cardStateInitialValues.goldImg)
        expect(props.cardState.isImgAvailable).toBe(cardStateInitialValues.isImgAvailable)
        expect(props.cardState.isOpen).toBe(cardStateInitialValues.isOpen)
    })

    it('fires swithImage onClick', () => {
        const switchButton = component.find('#switchImg').find(Button)
        expect(switchImage).not.toHaveBeenCalled()
        switchButton.simulate('click')
        expect(switchImage).toHaveBeenCalled()
    });

    it('fires addToFavorite onClick', () => {
        const switchButton = component.find('#favBtn').find(Button)
        expect(addToFavorite).not.toHaveBeenCalled()
        switchButton.simulate('click')
        expect(addToFavorite).toHaveBeenCalled()
    });
});
