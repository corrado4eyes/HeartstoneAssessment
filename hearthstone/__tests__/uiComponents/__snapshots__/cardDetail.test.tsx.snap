// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetail matches the snapshot 1`] = `
<CardDetail
  addToFavourites={[MockFunction]}
  card={
    Object {
      "artist": "Corrado Quattrocchi",
      "cardId": "what a dummy card!",
      "cardSet": "Dummy",
      "collectible": true,
      "cost": 9999,
      "faction": "Splendo",
      "flavor": "Meat",
      "howToGet": "Ask Frits",
      "howToGetGold": "Feeds and Walk Frits for 2 years every day (excluded the holidays)",
      "img": "http://media.services.zam.com/v1/media/byName/hs/cards/enus/DS1_183.png",
      "imgGold": "http://media.services.zam.com/v1/media/byName/hs/cards/enus/animated/DS1_183_premium.gif",
      "locale": "enUS",
      "name": "corrado4eyes",
      "playerClass": "Human",
      "rarity": "Super Rare",
      "text": "A dummy Italian boy",
      "type": "Developer",
    }
  }
  cardState={
    Object {
      "goldImg": false,
      "isImgAvailable": true,
      "isOpen": false,
    }
  }
  switchImage={[MockFunction]}
>
  <Container
    className="cardDetailContainer"
    fluid={false}
  >
    <div
      className="cardDetailContainer container"
    >
      <ForwardRef
        noGutters={false}
        style={
          Object {
            "height": "auto",
          }
        }
      >
        <div
          className="row"
          style={
            Object {
              "height": "auto",
            }
          }
        >
          <Col>
            <div
              className="col"
            >
              <Image
                fluid={false}
                height="90%"
                rounded={false}
                roundedCircle={false}
                src="http://media.services.zam.com/v1/media/byName/hs/cards/enus/DS1_183.png"
                thumbnail={false}
                width="90%"
              >
                <img
                  className=""
                  height="90%"
                  src="http://media.services.zam.com/v1/media/byName/hs/cards/enus/DS1_183.png"
                  width="90%"
                />
              </Image>
            </div>
          </Col>
          <Col
            className="cardDetail-font"
          >
            <div
              className="cardDetail-font col"
            >
              <Col>
                <div
                  className="col"
                >
                  corrado4eyes
                </div>
              </Col>
              <ForwardRef
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    >
                      Card set: 
                      Dummy
                    </div>
                  </Col>
                </div>
              </ForwardRef>
              <ForwardRef
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    >
                      Rarity: 
                      Super Rare
                    </div>
                  </Col>
                </div>
              </ForwardRef>
              <ForwardRef
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    >
                      Type: 
                      Developer
                    </div>
                  </Col>
                </div>
              </ForwardRef>
              <ForwardRef
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    >
                      Player class: 
                      Human
                    </div>
                  </Col>
                </div>
              </ForwardRef>
              <ForwardRef
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    >
                      Cost: 
                      9999
                       
                      <Image
                        fluid={false}
                        rounded={false}
                        roundedCircle={false}
                        src={Object {}}
                        thumbnail={false}
                      >
                        <img
                          className=""
                          src={Object {}}
                        />
                      </Image>
                    </div>
                  </Col>
                </div>
              </ForwardRef>
              <ForwardRef
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    />
                  </Col>
                  <Col>
                    <div
                      className="col"
                    />
                  </Col>
                </div>
              </ForwardRef>
            </div>
          </Col>
        </div>
      </ForwardRef>
      <ForwardRef
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col>
            <div
              className="col"
            >
              <Button
                active={false}
                className="no-borders"
                disabled={false}
                id="switchImg"
                onClick={[MockFunction]}
                type="button"
                variant="primary"
              >
                <button
                  className="no-borders btn btn-primary"
                  disabled={false}
                  id="switchImg"
                  onClick={[MockFunction]}
                  type="button"
                >
                  <Image
                    fluid={false}
                    rounded={false}
                    roundedCircle={false}
                    src={Object {}}
                    thumbnail={false}
                  >
                    <img
                      className=""
                      src={Object {}}
                    />
                  </Image>
                </button>
              </Button>
            </div>
          </Col>
          <Col
            sm="2"
          >
            <div
              className="col-sm-2"
            >
              <Button
                active={false}
                disabled={false}
                id="favBtn"
                onClick={[MockFunction]}
                type="button"
                variant="primary"
              >
                <button
                  className="btn btn-primary"
                  disabled={false}
                  id="favBtn"
                  onClick={[MockFunction]}
                  type="button"
                >
                  <Image
                    fluid={false}
                    rounded={false}
                    roundedCircle={false}
                    src={Object {}}
                    thumbnail={false}
                  >
                    <img
                      className=""
                      src={Object {}}
                    />
                  </Image>
                </button>
              </Button>
            </div>
          </Col>
        </div>
      </ForwardRef>
    </div>
  </Container>
</CardDetail>
`;
